src/context/EarnManagerProvider/providers/yearn/components/YearnManager/YearnDeposit.tsx

```diff
@@ -69,46 +149,102 @@ export const YearnDeposit = ({ api }: YearnDepositProps) => {
 
-  useEffect(() => {
-    ;(async () => {
-      if (!walletState.wallet) return
-      const chainAdapter = chainAdapterManager.byChain(ChainTypes.Ethereum)
-      setUserAddress(await chainAdapter.getAddress({ wallet: walletState.wallet }))
-    })()
-  }, [chain, chainAdapterManager, walletState])
-
   useEffect(() => {
     ;(async () => {
       try {
-        const apy = await api.apy({ vaultAddress })
-        setApy(apy)
+        if (!walletState.wallet) return
+        const [address, apy] = await Promise.all([
+          chainAdapter.getAddress({ wallet: walletState.wallet }),
+          api.apy({ vaultAddress })
+        ])
+        dispatch({ type: YearnActionType.SET_USER_ADDRESS, payload: address })
+        dispatch({ type: YearnActionType.SET_VAULT, payload: { apy } })
       } catch (error) {
         // TODO: handle client side errors
         console.error('error', error)
       }
     })()
-  }, [api, vaultAddress])
+  }, [api, chainAdapter, vaultAddress, walletState.wallet])
+
+  const getApproveEstimate = async () => {
+    if (!state.userAddress || !tokenId) return
+    const [gasLimit, gasPrice] = await Promise.all([
+      api.approveEstimatedGas({
+        spenderAddress: vaultAddress,
+        tokenContractAddress: tokenId,
+        userAddress: state.userAddress
+      }),
+      api.getGasPrice()
+    ])
+    return bnOrZero(gasPrice).times(gasLimit).toFixed(0)
+  }
 
   const handleContinue = async (formValues: DepositValues) => {
-    setValues(formValues)
-    if (!userAddress) return
+    if (!state.userAddress) return
+    // set deposit state for future use
+    dispatch({ type: YearnActionType.SET_DEPOSIT, payload: formValues })
+
+    // Check is approval is required for user address
     const _allowance = await api.allowance({
       tokenContractAddress: tokenId!,
       spenderAddress: vaultAddress,
-      userAddress
+      userAddress: state.userAddress
     })
     const allowance = bnOrZero(_allowance).div(`1e+${asset.precision}`)
-    allowance.gt(formValues.cryptoAmount)
-      ? memoryHistory.push(DepositPath.Confirm)
-      : memoryHistory.push(DepositPath.Approve)
+
+    // Skip approval step if user allowance is greater than requested deposit amount
+    if (allowance.gt(formValues.cryptoAmount)) {
+      memoryHistory.push(DepositPath.Confirm)
+    } else {
+      const estimatedGasCrypto = await getApproveEstimate()
+      if (!estimatedGasCrypto) return
+      dispatch({
+        type: YearnActionType.SET_APPROVE,
+        payload: { estimatedGasCrypto }
+      })
+      memoryHistory.push(DepositPath.Approve)
+    }
   }
 
   const handleApprove = async () => {
-    memoryHistory.push(DepositPath.Confirm)
+    if (!tokenId || !state.userAddress || !walletState.wallet) return
+    try {
+      dispatch({ type: YearnActionType.SET_LOADING, payload: true })
+      await api.approve({
+        bip32Params: {
```

Comment:
maybe use defaultBIP32Params from Ethereum chain adapter

Reply:
Yeah we tried that, but its not a valid type on the class